### PR TITLE
Task 4

### PR DESCRIPTION
**Pull request by achilove**
Main scope:

- [x] 1 - Task 4.1 is implemented
- [x] 3 - TASK 4.2 is implemented lambda links are provided and returns data
- [x] 4 - TASK 4.3 is implemented lambda links are provided and products is stored in DB (call TASK 4.2 to see the product)
- [x] 5 - Your own Frontend application is integrated with product service (/products API) and products from product-service are 
represented on Frontend. Link to a working Front-End application is provided for cross-check reviewer.

Additional (optional) tasks (but nice to have):

- [x] +1 - POST/products lambda functions returns error 400 status code if product data is invalid
- [x] +1 - All lambdas return error 500 status code on any error (DB connection, any unhandled error in code)
- [x] +1 - All lambdas do console.log for each incoming requests and their arguments
- [x] +1 - Transaction based creation of product 

endpoints:
  GET - https://33j0f1vd50.execute-api.eu-west-1.amazonaws.com/dev/products
  GET - https://33j0f1vd50.execute-api.eu-west-1.amazonaws.com/dev/products/{id}
  POST - https://33j0f1vd50.execute-api.eu-west-1.amazonaws.com/dev/products

FE link https://d3iuids07awyo3.cloudfront.net/
FE PR https://github.com/achilove/nodejs-aws-fe/pull/2

### CODE REVIEW

File: product-service/addProduct.js [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+            const stockInsertionResponse = await client.query(`insert into stocks(product_id, count) VALUES($1, $2) RETURNING *`, [product.id, count]);
+            await client.query('COMMIT')
+            product.count = stockInsertionResponse.rows[0].count;
+
+            return successResponse(product);
+        } catch (e) {
+            await client.query('ROLLBACK')
+
+            if((e.code && Number(e.code) === 23505) && (e.constraint && e.constraint == "products_title_key")  ){
+                return ProductExistsErrorResponse('product with given title already exists');
```
Comment (unfalse) [inline product-service/addProduct.js:57]: actually there was no condition to check if titles are the same in the homework

File: product-service/errors.js [hunk 2]
```diff
@@ -0,0 +1,16 @@
+class ValidationError extends Error {
+    constructor(param){
+        super();
+        this.message = param + " paramter value is invalid" ;
```
Comment (unfalse) [inline product-service/errors.js]: paramter -> parameter

File: product-service/getProduct.js [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+const { dbOptions } = require('./db/config');
+
+module.exports = async ( event ) => {
+    const client = new Client(dbOptions);
+    try {
+        await client.connect();
+        console.log('getProduct handler input:', event);
+        const { pathParameters } = event;
+        const { id } = pathParameters;
+        const { rows: products } = await client.query(`select products.id, products.title, products.price, products.description, stocks.count from products left join stocks on products.id = stocks.product_id where products.id='${id}';`);
```
Comment (unfalse) [inline product-service/getProduct.js]: you can make this query easier. For example: `select id, title, price, description, s.count from products p left join stocks s on p.id = s.product_id where p.id='${id}`

Comment (unfalse) [reply product-service/getProduct.js]: always use parameterized queries as it prevents possible sql injections https://node-postgres.com/features/queries

File: product-service/addProduct.js (hunk 1 above)
Comment (achilove) [reply product-service/addProduct.js:57]: yes, but it just common sense.

File: product-service/getProduct.js (hunk 3 above)
Comment (achilove) [reply product-service/getProduct.js]: you are right. I changed code to use parameterized quiery

**Pull request merged**